File: rxbinding-leanback-v17/src/main/java/com/jakewharton/rxbinding2/support/v17/leanback/widget/RxSearchBar.java
package com.jakewharton.rxbinding2.support.v17.leanback.widget;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.v17.leanback.widget.SearchBar;
import io.reactivex.Observable;
import io.reactivex.functions.Consumer;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkNotNull;

/**
 * Static factory methods for creating {@linkplain Observable observables} and {@linkplain Consumer
 * actions} for {@link SearchBar}.
 */
public final class RxSearchBar {
  /**
   * Create an observable of {@linkplain SearchBarSearchQueryEvent search query events} on {@code
   * view}.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   * <p>
   */
  @CheckResult
  @NonNull
  public static Observable<SearchBarSearchQueryEvent> searchQueryChangeEvents(
          @NonNull SearchBar view) {
    checkNotNull(view, "view == null");
    return new SearchBarSearchQueryChangeEventsOnSubscribe(view);
  }

  /**
   * Create an observable of String values for search query changes on {@code view}.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   * <p>
   */
  @CheckResult @NonNull
  public static Observable<String> searchQueryChanges(@NonNull SearchBar view) {
    checkNotNull(view, "view == null");
    return new SearchBarSearchQueryChangesOnSubscribe(view);
  }

  /**
   * An action which sets the searchQuery property of {@code view} with String values.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   *
   * @deprecated Use view::setSearchQuery method reference.
   */
  @Deprecated
  @CheckResult @NonNull
  public static Consumer<? super String> searchQuery(@NonNull SearchBar view) {
    checkNotNull(view, "view == null");
    return view::setSearchQuery;
  }

  private RxSearchBar() {
    throw new AssertionError("No instances.");
  }
}

File: rxbinding-leanback-v17-kotlin/src/main/kotlin/com/jakewharton/rxbinding2/support/v17/leanback/widget/RxSearchBar.kt
@file:Suppress(
    names = "NOTHING_TO_INLINE"
)

package com.jakewharton.rxbinding2.support.v17.leanback.widget

import android.support.annotation.CheckResult
import android.support.v17.leanback.widget.SearchBar
import io.reactivex.Observable
import io.reactivex.functions.Consumer

/**
 * Create an observable of {@linkplain SearchBarSearchQueryEvent search query events} on {@code
 * view}.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
@CheckResult
inline fun SearchBar.searchQueryChangeEvents(): Observable<SearchBarSearchQueryEvent> = RxSearchBar.searchQueryChangeEvents(this)

/**
 * Create an observable of String values for search query changes on `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
@CheckResult
inline fun SearchBar.searchQueryChanges(): Observable<String> = RxSearchBar.searchQueryChanges(this)

/**
 * An action which sets the searchQuery property of `view` with String values.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
@CheckResult
inline fun SearchBar.searchQuery(): Consumer<in String> = RxSearchBar.searchQuery(this)
